File: leetcode/graph/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/js/main.js
// Dijkstra
// Time O(E * ElogE)
// Space O(E^2)
const findTheCity = (n, edges, distanceThreshold) => {
  let adjList = [];
  let visited = new Set();
  let pq = new PriorityQueue({ comparator: (a, b) => a[1] - b[1] });

  for (let i = 0; i < n; i++) {
    adjList[i] = [];
  }

  for (let [u, v, weight] of edges) {
    adjList[u].push([v, weight]);
    adjList[v].push([u, weight]);
  }

  let min = Number.MAX_VALUE;
  let maxIndex = -Number.MAX_VALUE;

  for (let i = 0; i < n; i++) {
    let cntCities = dijkstra(i);

    if (cntCities <= min) {
      min = cntCities;
      maxIndex = i;
    }
  }

  return maxIndex;

  function dijkstra(u) {
    let cnt = 0;

    pq.offer([u, 0]);

    while (!pq.isEmpty()) {
      let [u, totalWeight] = pq.poll();

      if (visited.has(u) || totalWeight > distanceThreshold) continue;

      visited.add(u);
      cnt++;

      for (let [v, weight] of adjList[u]) {
        pq.offer([v, totalWeight + weight]);
      }
    }

    pq.clear();
    visited.clear();

    return cnt - 1;
  }
};

// Time O(N^2)
// Space O(N^2)
const findTheCity_II = (n, edges, distanceThreshold) => {
  let adjList = [];

  let distance = Array(n)
    .fill(0)
    .map(() => Array(n).fill(Number.MAX_VALUE));

  for (let i = 0; i < n; i++) {
    adjList[i] = [];
  }

  for (let [u, v, weight] of edges) {
    adjList[u].push([v, weight]);
    adjList[v].push([u, weight]);
  }

  let min = Number.MAX_VALUE;
  let maxIndex = -Number.MAX_VALUE;

  for (let i = 0; i < n; i++) {
    SPFA(i);
  }

  for (let i = 0; i < n; i++) {
    let cntCities = 0;

    for (let j = 0; j < n; j++) {
      if (i == j) continue;

      if (distance[i][j] <= distanceThreshold) cntCities++;
    }

    if (min >= cntCities) {
      min = cntCities;
      maxIndex = i;
    }
  }

  return maxIndex;

  function SPFA(i) {
    let queue = [];

    distance[i][i] = 0;
    queue.push(i);

    while (queue.length) {
      let size = queue.length;

      for (let z = 0; z < size; z++) {
        let u = queue.shift();

        for (let [v, dist] of adjList[u]) {
          if (distance[i][v] > distance[i][u] + dist) {
            distance[i][v] = distance[i][u] + dist;
            queue.push(v);
          }
        }
      }
    }
  }
};

File: leetcode/graph/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/py/main.py
import collections

class Solution:
    # DFS
    def findTheCity(self, n, edges, distanceThreshold):
        adjList = collections.defaultdict(list)
        maxCityIndex = -1
        cntCities = float('inf')
        distance = [[float('inf')] * n for _ in range(n)]

        def dfs(nodeId, u):
          for v, w in adjList[u]:
            if distance[nodeId][v] > distance[nodeId][u] + w and distance[nodeId][u] + w <= distanceThreshold:
              distance[nodeId][v] = distance[nodeId][u] + w
              dfs(nodeId, v)

        for u, v, w in edges:
          adjList[u].append([v, w])
          adjList[v].append([u, w])

        for u in range(n):
          distance[u][u] = 0          
          dfs(u, u)

        for i in range(n):
          cnt = 0

          for j in range(n):
            if i == j or distance[i][j] == float('inf'): continue
            cnt += 1

          if cntCities >= cnt:
            cntCities = cnt
            maxCityIndex = max(maxCityIndex, i)

        return maxCityIndex